### PR TITLE
LGA-3086: Setup CI/CD

### PR DESCRIPTION
**Pull request by said-moj**
## What does this pull request do?
Adds a CI/CD Pipeline using GitHub Actions' workflows

> [!IMPORTANT]  
> All ingress hosts are only accessible whilst using the Global Protect VPN.

## Feature Branch Pipeline
Screenshot 2024-05-19 at 14 40 31
All pushes to any branches, except main, trigger the feature branch pipeline, which performs the following jobs:

#### Static Analysis:
- Runs Ruff formatting

#### Testing:
- Runs Pytest
- Outputs results of Pytest in an artefact

#### Build:
- Runs `docker build` creating an image with a tag relating to the latest commit sha of the branch.
- Runs `docker push` pushing the image to the AWS Elastic Container Registry in the `laa-access-civil-legal-aid-production` namespace.

#### Deploy Dev:
- Installs or updates a Helm release, named after the branch.
- Requires only the static analysis step to pass.
- This creates a new release for each feature branch, with it's own deployment, service and ingress
- Ingress host is given by: `[branch name]-access-cla.cloud-platform.service.gov.uk`

#### Deploy UAT:
- Requires approval by a member of `laa-get-access` 
- Requires the static analysis and testing steps to pass
- Updates the Helm release on the namespace
- Ingress host is: [https://access-civil-legal-aid-uat.cloud-platform.service.justice.gov.uk/](https://access-civil-legal-aid-uat.cloud-platform.service.justice.gov.uk/)

## Release to staging & production
Screenshot 2024-05-19 at 15 16 51
Once a feature has been merged into main the release pipeline will run, performing the following jobs:

#### Static Analysis
- Runs Ruff formatting

#### Testing
- Runs Pytest
- Outputs results of Pytest in an artefact

#### Build
- Requires both the static analysis and testing steps to pass
- Uses docker to build and push the image to the ECR

#### Deploy Staging:
- Requires approval by a member of `laa-get-access` 
- Updates the Helm release on the namespace
- Ingress host is: [https://access-civil-legal-aid-staging.cloud-platform.service.justice.gov.uk/](https://access-civil-legal-aid-staging.cloud-platform.service.justice.gov.uk/)

#### Deploy Production:
- Requires approval by a member of `laa-get-access` 
- Updates the Helm release on the namespace
- Ingress host is: [https://access-civil-legal-aid.cloud-platform.service.justice.gov.uk/](https://access-civil-legal-aid.cloud-platform.service.justice.gov.uk/)

## Clean up dev release
Screenshot 2024-05-19 at 15 22 26

Once a Pull Request has been closed, or merged in main, the 'Clean up' workflow will run. This uninstalls the feature's Helm chart from the dev namespace, removing all deployments, ingresses and pods.

## Any other changes that would benefit highlighting?
Adds a Flask app which returns the text "Hello, World!" on the "/" route to test the deployment and ingress works as expected.

## Checklist

- [x] Provided JIRA ticket number in the title, e.g. "LGA-152: Sample title"

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+          KUBE_NAMESPACE: ${{ secrets.KUBE_NAMESPACE }}
+          KUBE_CLUSTER: ${{ secrets.KUBE_CLUSTER }}
+        run: |
+          echo "${{ secrets.KUBE_CERT }}" > ca.crt
+          kubectl config set-cluster ${KUBE_CLUSTER} --certificate-authority=./ca.crt --server=https://${KUBE_CLUSTER}
+          kubectl config set-credentials deploy-user --token=${{ secrets.KUBE_TOKEN }}
+          kubectl config set-context ${KUBE_CLUSTER} --cluster=${KUBE_CLUSTER} --user=deploy-user --namespace=${KUBE_NAMESPACE}
+          kubectl config use-context ${KUBE_CLUSTER}
+
+      - name: Assume role in Cloud Platform
```
Comment (digitalali-moj) [inline .github/workflows/deploy.yml:50]: Might be worth checking these two bits, I think I normally add ECR registry as variable and not require login for deployment.

File: helm_deploy/laa-access-civil-legal-aid/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+              path: /
+              port: http
+          readinessProbe:
+            httpGet:
+              path: /
+              port: http
+          resources:
+            {{- toYaml .Values.resources | nindent 12 }}
+          {{- with .Values.volumeMounts }}
+          volumeMounts:
```
Comment (digitalali-moj) [inline helm_deploy/laa-access-civil-legal-aid/templates/deployment.yaml:54]: Do we need this?

File: .github/workflows/deploy.yml (hunk 1 above)
Comment (BenMillar-CGI) [reply .github/workflows/deploy.yml]: Thanks Jav, great point! I have moved the ECR Registry URL to a repository secret with this commit: [https://github.com/ministryofjustice/laa-access-civil-legal-aid/pull/3/commits/9024ed52c39ce0f99ad8be4fba416db987efe8af](https://github.com/ministryofjustice/laa-access-civil-legal-aid/pull/3/commits/9024ed52c39ce0f99ad8be4fba416db987efe8af)

File: helm_deploy/laa-access-civil-legal-aid/templates/deployment.yaml (hunk 2 above)
Comment (BenMillar-CGI) [reply helm_deploy/laa-access-civil-legal-aid/templates/deployment.yaml]: Hi Jav, yes, I agree! I have removed a lot of the unused boiler plate with this PR here: [https://github.com/ministryofjustice/laa-access-civil-legal-aid/pull/3/commits/daebd24335c5d6a1c634d5cfcd11ece2b4034bd3](https://github.com/ministryofjustice/laa-access-civil-legal-aid/pull/3/commits/daebd24335c5d6a1c634d5cfcd11ece2b4034bd3)

Many thanks! 😄

**Review by Pheonnexx — APPROVED** [review]
LGTM good catch by Jav on somethings!

**Review by digitalali-moj — APPROVED** [review]
Legend.

**Pull request merged**